File: AdaptiveTD/AdaptiveTD/AdaptiveTD/Enemy.cs
using System;
using System.Collections.Generic;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace AdaptiveTD
{
    class Enemy
    {
        Texture2D enemyTexture, healthBarRedTexture, healthBarYellowTexture;
        int health, maxHealth;
        public int Health
        {
            get { return health; }
            set { health = value; }
        }
        int normalSpeed, currentSpeed;
        int goldYield;
        public int GoldYield
        {
            get { return goldYield; }
        }
        Vector2 targetPosition;
        Vector2 position;
        public Vector2 Position
        {
            get { return position; }
            set { position = value; }
        }
        List<Direction> directions = new List<Direction>();
        float rotation;
        Direction currentDirection = Direction.None;
        Vector2 origin;
        public Vector2 Origin
        {
            get { return origin; }
        }
        float distanceTravelled;
        public float DistanceTravelled
        {
            get { return distanceTravelled; }
        }
        int directionCounter = 0;
        Rectangle healthBarYellowRectangle, healthBarRedRectangle;
        Slow slow;
        public Slow Slow
        {
            get { return slow; }
            set { slow = value; }
        }
        DamageOverTime damageOverTime;
        public DamageOverTime DamageOverTime
        {
            get { return damageOverTime; }
            set { damageOverTime = value; }
        }
        Color color;
        public Color Color
        {
            get { return color; }
            set { color = value; }
        }
        
        public Enemy(Vector2 startPosition, EnemyInfo enemyInfo, List<Direction> directions)
        {
            this.enemyTexture = enemyInfo.EnemyTexture;
            this.healthBarRedTexture = enemyInfo.RedHealthBar;
            this.healthBarYellowTexture = enemyInfo.YellowHealthBar;
            origin = new Vector2(enemyTexture.Width / 2, enemyTexture.Height / 2);
            this.position = new Vector2(startPosition.X * GameConstants.tileSize, startPosition.Y * GameConstants.tileSize);
            targetPosition = position;
            this.health = enemyInfo.Health;
            this.maxHealth = health;
            this.normalSpeed = enemyInfo.Speed;
            this.currentSpeed = normalSpeed;
            this.goldYield = enemyInfo.GoldYield;
            this.directions = directions;
            color = Color.White;
            slow = new Slow(false);
            damageOverTime = new DamageOverTime(false);
        }

        public void Turn(bool right)
        {
            if (right)
                rotation -= (float)(Math.PI / 2);
            else
                rotation += (float)(Math.PI / 2);
        }

        public virtual void OnDeath()
        {

        }

        public void Update(float gameTime)
        {
            if (currentDirection == Direction.None)
            {
                if (directionCounter < directions.Count)
                {
                    currentDirection = directions[directionCounter];
                    directionCounter++;
                }
                else
                    currentDirection = Direction.Right;
                switch (currentDirection)
                {
                    case Direction.Up:
                        targetPosition += new Vector2(0, -GameConstants.tileSize);
                        break;
                    case Direction.Down:
                        targetPosition += new Vector2(0, GameConstants.tileSize);
                        break;
                    case Direction.Left:
                        targetPosition += new Vector2(-GameConstants.tileSize, 0);
                        break;
                    case Direction.Right:
                        targetPosition += new Vector2(GameConstants.tileSize, 0);
                        break;
                }
            }

            switch (currentDirection)
            {
                case Direction.Up:
                    position.Y -= (float)(currentSpeed * gameTime * slow.Percentage / 100);
                    break;
                case Direction.Down:
                    position.Y += (float)(currentSpeed * gameTime * slow.Percentage / 100);
                    break;
                case Direction.Left:
                    position.X -= (float)(currentSpeed * gameTime * slow.Percentage / 100);
                    break;
                case Direction.Right:
                    position.X += (float)(currentSpeed * gameTime * slow.Percentage / 100);
                    break;
            }

            distanceTravelled += (float)(currentSpeed * gameTime);

            if (Math.Abs(position.X - targetPosition.X) < 1 && Math.Abs(position.Y - targetPosition.Y) < 1)
            {
                position = targetPosition;
                currentDirection = Direction.None;
            }

            damageOverTime.DurationSinceLastTick -= gameTime;
            if (damageOverTime.DurationSinceLastTick <= 0)
            {
                damageOverTime.DurationSinceLastTick = damageOverTime.Duration / damageOverTime.Ticks;
                health -= damageOverTime.DamagePerTick;
            }
            damageOverTime.RemainingDuration -= gameTime;
            if (damageOverTime.RemainingDuration <= 0)
            {
                damageOverTime = new DamageOverTime(false);
                color = Color.White;
            }
            else
            {
                color = Color.LightBlue;
            }

            slow.Duration -= gameTime;
            if (slow.Duration <= 0)
            {
                slow = new Slow(false);
                color = Color.White;
            }
            else
            {
                color = Color.LightBlue;
            }
            healthBarRedRectangle = new Rectangle((int)position.X, (int)position.Y - 10, 64, 5);
            healthBarYellowRectangle = new Rectangle((int)position.X, (int)position.Y - 10, (int)((float)64 * (float)health / (float)maxHealth), 5);
        }

        public void Draw(SpriteBatch spriteBatch)
        {
            Vector2 drawPosition = position + origin;
            spriteBatch.Draw(enemyTexture, drawPosition, null, color, rotation, origin, 1.0f, SpriteEffects.None, 1.0f);
            spriteBatch.Draw(healthBarRedTexture, healthBarRedRectangle, Color.White);
            spriteBatch.Draw(healthBarYellowTexture, healthBarYellowRectangle, Color.White);
            
        }

    }

    public struct EnemyInfo
    {
        int health;

        public int Health
        {
            get { return health; }
            set { health = value; }
        }
        int speed;

        public int Speed
        {
            get { return speed; }
            set { speed = value; }
        }

        int goldYield;

        public int GoldYield
        {
            get { return goldYield; }
            set { goldYield = value; }
        }

        Texture2D redHealthBar;

        public Texture2D RedHealthBar
        {
            get { return redHealthBar; }
            set { redHealthBar = value; }
        }
        Texture2D yellowHealthBar;

        public Texture2D YellowHealthBar
        {
            get { return yellowHealthBar; }
            set { yellowHealthBar = value; }
        }
        Texture2D enemyTexture;

        public Texture2D EnemyTexture
        {
            get { return enemyTexture; }
            set { enemyTexture = value; }
        }

        public EnemyInfo(int health, int speed, int goldYield, Texture2D enemyTexture, Texture2D redHealthBar, Texture2D yellowHealthBar)
        {
            this.health = health;
            this.speed = speed;
            this.goldYield = goldYield;
            this.enemyTexture = enemyTexture;
            this.yellowHealthBar = yellowHealthBar;
            this.redHealthBar = redHealthBar;
        }
    }
}

File: LibGDXProjects/AdaptiveTD/src/no/uia/adaptiveTD/Enemy.java
package no.uia.adaptiveTD;

import java.util.HashMap;
import java.util.List;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;


public class Enemy {
    EnemyStats enemyStats;
	int currentHealth, currentSpeed, directionCounter; 
    Vector2 position, targetPosition, origin;
    float rotation, distanceTravelled;
    Rectangle healthBarYellowRectangle, healthBarRedRectangle;
    List<Direction> directions;
    Direction currentDirection = Direction.None;
	HashMap<Integer, Sprite> sprites;
    
    public Enemy(Vector2 startPosition, EnemyStats enemyStats, List<Direction> directions, Sprite enemySprite, Sprite redHealthBarSprite, Sprite yellowHealthBarSprite)
    {
    	this.enemyStats = enemyStats;
    	this.directions = directions;
        origin = new Vector2(GameConstants.tileSize / 2, GameConstants.tileSize / 2);
        this.position = new Vector2(startPosition.x * GameConstants.tileSize, startPosition.y * GameConstants.tileSize);
        targetPosition = new Vector2(position.x, position.y);
        this.currentHealth = enemyStats.getHealth();
        this.currentSpeed =  enemyStats.getSpeed();
        this.sprites = new HashMap<Integer, Sprite>();
        sprites.put(0, enemySprite);
        sprites.put(1, redHealthBarSprite);
        sprites.put(2, yellowHealthBarSprite);
    }
    public void Turn(Boolean right)
    {
        if (right)
            rotation -= (float)(Math.PI / 2);
        else
            rotation += (float)(Math.PI / 2);
    }

    public void Update(float gameTime)
    {
    	System.out.println(gameTime);
        if (currentDirection == Direction.None)
        {
            if (directionCounter < directions.size())
            {
                currentDirection = directions.get(directionCounter);
                directionCounter++;
            }
            else
                currentDirection = Direction.Right;
            switch (currentDirection)
            {
                case Up:
                    targetPosition.y -= GameConstants.tileSize;
                    break;
                case Down:
                    targetPosition.y += GameConstants.tileSize;
                    break;
                case Left:
                    targetPosition.x -= GameConstants.tileSize;
                    break;
                case Right:
                    targetPosition.x += GameConstants.tileSize;
                    break;
                default:
                	break;
            }
        }

        switch (currentDirection)
        {
            case Up:
                //position.y -= (float)(currentSpeed * gameTime - currentSpeed * gameTime * slow.Percentage / 100);
            	position.y -= (float)(currentSpeed * gameTime);
                break;
            case Down:
                position.y += (float)(currentSpeed * gameTime);
                break;
            case Left:
                position.x -= (float)(currentSpeed * gameTime);
                break;
            case Right:
                position.x += (float)(currentSpeed * gameTime);
                break;
            default:
            	break;
        }
//distanceTravelled += (float)(currentSpeed * gameTime - currentSpeed * gameTime * slow.Percentage / 100);
        distanceTravelled += (float)(currentSpeed * gameTime - currentSpeed * gameTime);

        if (Math.abs(position.x - targetPosition.x) < 1 && Math.abs(position.y - targetPosition.y) < 1)
        {
            position = new Vector2(targetPosition.x, targetPosition.y);
            currentDirection = Direction.None;
        }

        /*
        damageOverTime.DurationSinceLastTick -= gameTime;
        if (damageOverTime.DurationSinceLastTick <= 0)
        {
            damageOverTime.DurationSinceLastTick = damageOverTime.Duration / damageOverTime.Ticks;
            health -= damageOverTime.DamagePerTick;
        }
        Boolean isSlowed, isDotted, isAffected;
        damageOverTime.RemainingDuration -= gameTime;
        if (damageOverTime.RemainingDuration <= 0)
        {
            damageOverTime = new DamageOverTime(false);
            isDotted = false;
        }
        else
        {
            isDotted = true;
        }

        slow.Duration -= gameTime;
        if (slow.Duration <= 0)
        {
            slow = new Slow(false);
            isSlowed = false;
        }
        else
        {
            isSlowed = true;
        }
        isAffected = isSlowed || isDotted;
        if (isAffected)
            color = Color.LightGreen;
        else
            color = Color.White;
        */
        healthBarRedRectangle = new Rectangle((int)position.x, (int)position.y - 10, 64, 5);
        healthBarYellowRectangle = new Rectangle((int)position.x, (int)position.y - 10, (int)((float)64 * (float)currentHealth / (float)enemyStats.getHealth()), 5);
    }

    public void Draw(SpriteBatch spriteBatch)
    {
        Vector2 drawPosition = new Vector2(position.x + origin.x, position.y + origin.y);
        spriteBatch.draw(sprites.get(0), position.x, GameConstants.morphY(position.y));
        spriteBatch.draw(sprites.get(1), healthBarRedRectangle.getX(), GameConstants.morphY(healthBarRedRectangle.getY() - GameConstants.tileSize), healthBarRedRectangle.getWidth(), healthBarRedRectangle.getHeight());
        spriteBatch.draw(sprites.get(2), healthBarYellowRectangle.getX(), GameConstants.morphY(healthBarYellowRectangle.getY() - GameConstants.tileSize), healthBarYellowRectangle.getWidth(), healthBarYellowRectangle.getHeight());
        
        /*
        spriteBatch.Draw(enemyTexture, drawPosition, null, color, rotation, origin, 1.0f, SpriteEffects.None, 1.0f);
        spriteBatch.Draw(redHealthBar, healthBarRedRectangle, Color.White);
        spriteBatch.Draw(yellowHealthBar, healthBarYellowRectangle, Color.White);
        */
    }
	public EnemyStats getEnemyInfo() {
		return enemyStats;
	}
	public void setEnemyInfo(EnemyStats enemyInfo) {
		this.enemyStats = enemyInfo;
	}
	public int getCurrentHealth() {
		return currentHealth;
	}
	public void setCurrentHealth(int currentHealth) {
		this.currentHealth = currentHealth;
	}
	public int getCurrentSpeed() {
		return currentSpeed;
	}
	public void setCurrentSpeed(int currentSpeed) {
		this.currentSpeed = currentSpeed;
	}
	public int getDirectionCounter() {
		return directionCounter;
	}
	public void setDirectionCounter(int directionCounter) {
		this.directionCounter = directionCounter;
	}
	public Vector2 getPosition() {
		return position;
	}
	public void setPosition(Vector2 position) {
		this.position = position;
	}
	public Vector2 getTargetPosition() {
		return targetPosition;
	}
	public void setTargetPosition(Vector2 targetPosition) {
		this.targetPosition = targetPosition;
	}
	public Vector2 getOrigin() {
		return origin;
	}
	public void setOrigin(Vector2 origin) {
		this.origin = origin;
	}
	public float getRotation() {
		return rotation;
	}
	public void setRotation(float rotation) {
		this.rotation = rotation;
	}
	public float getDistanceTravelled() {
		return distanceTravelled;
	}
	public void setDistanceTravelled(float distanceTravelled) {
		this.distanceTravelled = distanceTravelled;
	}
	public Rectangle getHealthBarYellowRectangle() {
		return healthBarYellowRectangle;
	}
	public void setHealthBarYellowRectangle(Rectangle healthBarYellowRectangle) {
		this.healthBarYellowRectangle = healthBarYellowRectangle;
	}
	public Rectangle getHealthBarRedRectangle() {
		return healthBarRedRectangle;
	}
	public void setHealthBarRedRectangle(Rectangle healthBarRedRectangle) {
		this.healthBarRedRectangle = healthBarRedRectangle;
	}
	public List<Direction> getDirections() {
		return directions;
	}
	public void setDirections(List<Direction> directions) {
		this.directions = directions;
	}
	public Direction getCurrentDirection() {
		return currentDirection;
	}
	public void setCurrentDirection(Direction currentDirection) {
		this.currentDirection = currentDirection;
	}
	public HashMap<Integer, Sprite> getSprites() {
		return sprites;
	}
	public void setSprites(HashMap<Integer, Sprite> sprites) {
		this.sprites = sprites;
	}

}
